### ISSUE TITLE
device: udev rule SYSTEMD_WANTS does not particular work for service with time consuming ExecStop= defined

### ISSUE THREAD

**Issue by Shuangistan** [systemd/systemd#11526]
**systemd version the issue has been seen with**
> 237

**Expected behaviour you didn't see**
 > Remove and Reconnect the USB device, the wanted service shall be started.

**Unexpected behaviour you saw**
 > Service does not start.

**Steps to reproduce the problem**

1. For the USB device, define `TAG+="systemd", ENV{SYSTEMD_WANTS}+="device@%k.service"`.
2. In `device@.service`, `BindsTo=device-%i.device` to the USB device unit.
3. Define e.g. `ExecStop=/bin/sleep 5` in the service file.

4. Disconnect and connect the device:
If within 5s, `device@.service` will not start.
Otherwise, `device@.service` starts as expected.

During the ExecStop, the service is in status
`Active: deactivating (stop) since Mon 2018-12-10 13:40:23 UTC; 2s ago`
In case the device is connected at this time, it will not start and remain inactive
`Active: inactive (dead) since Mon 2018-12-10 13:40:27 UTC; 11s ago`

This issue is mentioned and discussed in #11305 and
https://lists.freedesktop.org/archives/systemd-devel/2019-January/041942.html
But it has not been fixed.

**Comment by bl33pbl0p** [comment]
Just for the record, the "Why" behind this is same as the linked bug, the start job is triggered by the device units code with JOB_FAIL job mode, but a conflicting job type is already installed on the unit and running (stop), hence the transaction is deemed destructive and fails.

Most likely, replacing it with JOB_REPLACE should make things work, however it has to be seen if that would break anything (again, most probably not).

**Comment by Shuangistan** [comment]
I can confirm your idea. Change with the following will solve my issue.
After 5 seconds sleep, the service starts again.

However, I am checking if it breaks.

```diff
diff --git a/src/core/unit.c b/src/core/unit.c
index 24b14fb..549604c 100644
--- a/src/core/unit.c
+++ b/src/core/unit.c
@@ -2070,7 +2070,7 @@ static void retroactively_start_dependencies(Unit *u) {
         HASHMAP_FOREACH_KEY(v, other, u->dependencies[UNIT_WANTS], i)
                 if (!hashmap_get(u->dependencies[UNIT_AFTER], other) &&
                     !UNIT_IS_ACTIVE_OR_ACTIVATING(unit_active_state(other)))
-                        manager_add_job(u->manager, JOB_START, other, JOB_FAIL, NULL, NULL);
+                        manager_add_job(u->manager, JOB_START, other, JOB_REPLACE, NULL, NULL);
 
         HASHMAP_FOREACH_KEY(v, other, u->dependencies[UNIT_CONFLICTS], i)
                 if (!UNIT_IS_INACTIVE_OR_DEACTIVATING(unit_active_state(other)))
-- 
2.7.4
```

**Comment by bl33pbl0p** [comment]
@Shuangistan Yes, that's the correct place to change this. You may also need to replace the JOB_FAIL with JOB_REPLACE in device.c (so that if you add SYSTEMD_WANTS after device has been plugged, and start jobs are triggered, they still execute even if the said unit is deactivating by waiting for the service to stop).

One problem here is that this replacement works reliably only in cases of services. Start jobs keeps waiting until deactivation triggered by some prior stop job it is replacing completes. Now, in that time if you again trigger stop job then it will also not start the unit, instead just make it sit in inactive state (by virtue of job, as it satisfies what the intended state to be reached is). Some units (primarily devices, but also targets) use jobs to encapsulate (or enclose) a unit specific operation, hence replacing the job also ends up cancelling the effect the job was going to have. For example if you replace a device's start job with a stop job, then it will not wait until 1min 30s (or whatever default timeout is) for device to appear before making stop job runnable. Instead, it immediately stops waiting for device appear and stop job is run, waiting for it to go away (which is already the case).

So yes, JOB_REPLACE might look like the correct solution here (and is for services), it most certainly isn't for other unit types for aforementioned reasons. This is a little in the area of design inconsistency, jobs having multiple meanings and effects on cancellation depending on unit type.

**Comment by bl33pbl0p** [comment]
This is also why I did not make that change so far, I think there needs to be some clear establishment of what jobs are supposed to do in some generalized manner before this can be tackled.

**Comment by Shuangistan** [comment]
@bl33pbl0p Thanks for the detailed information. I wasn't thinking into the direction of other types of units (.target, .device). I see the for Requires=, the similar thing has been implemented
```c
        HASHMAP_FOREACH_KEY(v, other, u->dependencies[UNIT_REQUIRES], i)
                if (!hashmap_get(u->dependencies[UNIT_AFTER], other) &&
                    !UNIT_IS_ACTIVE_OR_ACTIVATING(unit_active_state(other)))
                        manager_add_job(u->manager, JOB_START, other, JOB_REPLACE, NULL, NULL);
```
So I am looking into the direction of the difference between Requires= and Wants=.
Thanks for pointing out.

My other thought was, as this use case is specially in use with udev. Maybe a stronger dependency introducing in the udev rule would help. Again, the effects are not checked.

**Comment by Shuangistan** [comment]
@bl33pbl0p Not sure if we could to discuss for all the 11 unit types. How about to add another special hack to check if the UNIT_TYPE is UNIT_SERVICE and only change the aforementioned lines to JOB_REPLACE for UNIT_SERVICE (maybe also UNIT_SOCKET etc.)?
At least I could give it a try at my downstream.
